x-pack/legacy/plugins/ingest/server/libs/adapters/so_database/mocked_adapter.ts

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Slapshot from '@mattapperson/slapshot';
+import {
+  SavedObjectAttributes,
+  SavedObjectsBulkCreateObject,
+  SavedObjectsBaseOptions,
+  SavedObjectsFindOptions,
+  SavedObjectsFindResponse,
+  SavedObjectsBulkResponse,
+  SavedObject,
+  SavedObjectsUpdateOptions,
+} from 'src/core/server';
+import {
+  SavedObjectsCreateOptions,
+  SavedObjectsBulkGetObject,
+  SavedObjectsUpdateResponse,
+} from 'target/types/server';
+import { SODatabaseAdapter } from './default';
+
+const { memorize } = Slapshot;
+
+/**
+ * MockedSoDatabaseAdapter that use Slapshot
+ */
+export class MockedSODatabaseAdapter implements Omit<SODatabaseAdapter, 'client'> {
+  constructor(private readonly soAdapter?: SODatabaseAdapter) {}
+  /**
+   * Persists a SavedObject
+   *
+   * @param type
+   * @param attributes
+   * @param options
+   */
+  async create<T extends SavedObjectAttributes = any>(
+    type: string,
+    data: T,
+    options?: SavedObjectsCreateOptions
+  ) {
+    return memorize(
+      `create:${type}:${JSON.stringify(data)}:${JSON.stringify(options || {})}`,
+      async () => {
+        if (!this.soAdapter) {
+          throw new Error('You need to provide a valid saved object adapter');
+        }
+
+        return this.soAdapter.create(type, data, options);
+      }
+    );
+  }
+
+  /**
+   * Persists multiple documents batched together as a single request
+   *
+   * @param objects
+   * @param options
+   */
+  async bulkCreate<T extends SavedObjectAttributes = any>(
+    objects: Array<SavedObjectsBulkCreateObject<T>>,
+    options?: SavedObjectsCreateOptions
+  ) {
+    return memorize(`bulk_create:${objects}:${JSON.stringify(options)}`, async () => {
+      if (!this.soAdapter) {
+        throw new Error('You need to provide a valid saved object adapter');
+      }
+
+      return this.soAdapter.bulkCreate(objects, options);
+    });
+  }
+
+  /**
+   * Deletes a SavedObject
+   *
+   * @param type
+   * @param id
+   * @param options
+   */
+  async delete(type: string, id: string, options: SavedObjectsBaseOptions = {}) {
+    return memorize(`delete:${type}:${id}:${JSON.stringify(options)}`, async () => {
```

Comment:
Please update these and other `return <promise>` statements to use `return await`, which enables better async stack traces and makes things a little more explicit.